libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,637 @@
+  typename __concat_value_t<_Rs...>;
+  typename __concat_rvalue_reference_t<_Rs...>;
+} && __concat_indirectly_readable<_Rs...>;
+
+template <bool _Const, class... _Rs>
+concept __concat_is_random_access =
+    (random_access_range<__maybe_const<_Const, _Rs>> && ...) && (sized_range<__maybe_const<_Const, _Rs>> && ...);
+
+template <bool _Const, class... _Rs>
+concept __concat_is_bidirectional =
+    ((bidirectional_range<__maybe_const<_Const, _Rs>> && ...) && (common_range<__maybe_const<_Const, _Rs>> && ...));
+
+template <bool _Const, class... _Views>
+concept __all_forward = (forward_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class... _Tp>
+struct __apply_drop_first;
+
+template <bool _Const, class _Head, class... _Tail>
+struct __apply_drop_first<_Const, _Head, _Tail...> {
+  static constexpr bool value = (sized_range<__maybe_const<_Const, _Tail>> && ...);
+};
+
+template <input_range... _Views>
+  requires(view<_Views> && ...) && (sizeof...(_Views) > 0) && __concatable<_Views...>
+class concat_view : public view_interface<concat_view<_Views...>> {
+  tuple<_Views...> __views_;
+
+  template <bool _Const>
+  class __iterator;
+
+public:
+  _LIBCPP_HIDE_FROM_ABI constexpr concat_view() = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit concat_view(_Views... __views) : __views_(std::move(__views)...) {}
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator<false> begin()
+    requires(!(__simple_view<_Views> && ...))
+  {
+    __iterator<false> __it(this, in_place_index<0>, ranges::begin(std::get<0>(__views_)));
+    __it.template __satisfy<0>();
+    return __it;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator<true> begin() const
+    requires((range<const _Views> && ...) && __concatable<const _Views...>)
+  {
+    __iterator<true> __it(this, in_place_index<0>, ranges::begin(std::get<0>(__views_)));
+    __it.template __satisfy<0>();
+    return __it;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto end()
+    requires(!(__simple_view<_Views> && ...))
+  {
+    if constexpr (common_range<typename __last_view<_Views...>::type>) {
+      constexpr auto __n = sizeof...(_Views);
+      return __iterator<false>(this, in_place_index<__n - 1>, ranges::end(std::get<__n - 1>(__views_)));
+    } else {
+      return default_sentinel;
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto end() const
+    requires(range<const _Views> && ...)
+  {
+    if constexpr (common_range<typename __last_view<_Views...>::type>) {
+      constexpr auto __n = sizeof...(_Views);
+      return __iterator<true>(this, in_place_index<__n - 1>, ranges::end(std::get<__n - 1>(__views_)));
+    } else {
+      return default_sentinel;
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto size()
+    requires(sized_range<_Views> && ...)
+  {
+    return std::apply(
+        [](auto... __sizes) {
+          using CT = make_unsigned_t<common_type_t<decltype(__sizes)...>>;
+          return (CT(__sizes) + ...);
```

Comment:
Missed __uglification. But given `CT` is used only once, I think we can avoid introducing this type alias, ditto below and for `reference` in `__iterator::operator*()`.